implementations/rust/ockam/ockam_core/src/error/inner/mod.rs

```diff
@@ -0,0 +1,183 @@
+    #[cfg(feature = "std")]
+    payload: Vec<PayloadEntry>,
+    #[cfg(feature = "std")]
+    source_loc: Location,
+    #[cfg(feature = "std")]
+    #[serde(skip, default)]
+    cause: Option<BoxDynErr>,
+    #[cfg(feature = "std")]
+    #[serde(skip, default)]
+    local: Option<Vec<LocalPayloadEntry>>,
+}
+
+impl ErrorData {
+    #[cold]
+    #[track_caller]
+    #[cfg(feature = "std")]
+    pub(super) fn new<E>(code: ErrorCode, cause: E) -> Self
+    where
+        E: Into<Box<dyn ErrorTrait + Send + Sync>>,
+    {
+        Self::new_inner(code, Some(cause.into()), Some(core::any::type_name::<E>()))
+    }
+    #[cold]
+    #[track_caller]
+    #[cfg(feature = "std")]
+    pub(super) fn new_without_cause(origin: Origin, kind: Kind) -> Self {
+        Self::new_inner(ErrorCode::new(origin, kind), None, None)
+    }
+
+    // FIXME
+    #[cold]
+    #[track_caller]
+    #[cfg(not(feature = "std"))]
+    pub(super) fn new<E>(code: ErrorCode, _cause: E) -> Self {
+        Self { code }
+    }
+
+    #[cold]
+    #[track_caller]
+    #[cfg(not(feature = "std"))]
+    pub(super) fn new_without_cause(origin: Origin, kind: Kind) -> Self {
+        Self {
+            code: ErrorCode::new(origin, kind),
+        }
+        // Self::new_inner(ErrorCode::new(origin, kind), None, None)
+    }
+
+    #[cold]
+    #[track_caller]
+    #[cfg(feature = "std")]
+    pub(super) fn new_inner(
+        code: ErrorCode,
+        cause: Option<Box<dyn ErrorTrait + Send + Sync>>,
+        type_name: Option<&'static str>,
+    ) -> Self {
+        let location = core::panic::Location::caller();
+        #[allow(unused_mut)]
+        let mut local: Vec<LocalPayloadEntry> = vec![];
+
+        #[cfg(all(feature = "std", feature = "error-traces"))]
+        if trace_config::BACKTRACE_ENABLED.get() {
+            local.push(LocalPayloadEntry::Backtrace(backtrace::Backtrace::new()));
+        }
+
+        #[cfg(all(feature = "std", feature = "error-traces"))]
+        if trace_config::SPANTRACE_ENABLED.get() {
+            local.push(LocalPayloadEntry::Spantrace(
+                tracing_error::SpanTrace::capture(),
+            ));
+        }
+
+        let mut payload = vec![];
+        if let Some(cause) = cause.as_ref() {
+            let debug = crate::compat::format!("{:?}", cause);
+            let display = crate::compat::format!("{}", cause);
+            payload.push(PayloadEntry::Cause {
+                display,
+                debug,
+                type_name: type_name.unwrap_or("<unknown>").to_owned(),
+            });
+        }
+
+        Self {
+            code,
+            cause,
+            source_loc: location.into(),
+            local: Some(local),
+            payload,
+        }
+    }
+
+    #[cold]
+    pub fn add_context(&mut self, _key: &str, _val: &dyn core::fmt::Display) {
+        #[cfg(feature = "std")]
+        self.payload.push(PayloadEntry::Info(
```

Comment:
Currently, yes. But it would make downcasting impossible, and loose Debug formatting, unlike use of the Error trait. I would like people who implement Worker to be able to get their errors out again on the other side, at least on std platforms.

The only difficulty here is due to a constraint that existed when this code was initially written (round-trip serialization of errors) that no longer exists.

Reply:
Ah, wait, `add_context` yes that's fine, but I have a further rework for this. (But it will happen as a follow-up, since this is already an enormous PR thats too fragile to keep in a branch for much longer).